services/streammanager/kafka/kafkamanager.go

```diff
@@ -354,86 +399,75 @@ func Produce(jsonData json.RawMessage, producer interface{}, destConfig interfac
 	var batch []map[string]interface{}
 	err := json.Unmarshal(jsonData, &batch)
 	if err != nil {
-		return 400, "Failure", "Error while unmarshalling json data :: " + err.Error()
+		return 400, "Failure", "Error while unmarshalling json data: " + err.Error()
 	}
 
-	batchedMessage, err := prepareBatchedMessage(topic, batch, timestamp)
+	timestamp := time.Now()
+	batchOfMessages, err := prepareBatchOfMessages(topic, batch, timestamp)
 	if err != nil {
-		return 400, "Failure", "Error while preparing batched message :: " + err.Error()
+		return 400, "Failure", "Error while preparing batched message: " + err.Error()
 	}
 
-	err = kafkaProducer.SendMessages(batchedMessage)
+	err = p.Publish(ctx, batchOfMessages...)
 	if err != nil {
 		return makeErrorResponse(err) // would retry the messages in batch in case brokers are down
 	}
 
 	returnMessage := "Kafka: Message delivered in batch"
-	statusCode := 200
-	errorMessage := returnMessage
-
-	return statusCode, returnMessage, errorMessage
+	return 200, returnMessage, returnMessage
 }
 
-func sendMessage(jsonData json.RawMessage, kafkaProducer sarama.SyncProducer, topic string) (int, string, string) {
-	timestamp := time.Now()
+func sendMessage(ctx context.Context, jsonData json.RawMessage, p producer, topic string) (int, string, string) {
 	parsedJSON := gjson.ParseBytes(jsonData)
-	data := parsedJSON.Get("message").Value().(interface{})
+	messageValue := parsedJSON.Get("message").Value()
+	if messageValue == nil {
+		return 400, "Failure", "Invalid message"
+	}
+
+	data := messageValue.(interface{})
 	value, err := json.Marshal(data)
 	if err != nil {
 		return makeErrorResponse(err)
 	}
-	userID, _ := parsedJSON.Get("userId").Value().(string)
 
+	timestamp := time.Now()
+	userID, _ := parsedJSON.Get("userId").Value().(string)
 	message := prepareMessage(topic, userID, value, timestamp)
-
-	partition, offset, err := kafkaProducer.SendMessage(message)
-	if err != nil {
+	if err = p.Publish(ctx, message); err != nil {
 		return makeErrorResponse(err)
 	}
 
-	returnMessage := fmt.Sprintf("Message delivered at Offset: %v , Partition: %v for topic: %s", offset, partition, topic)
-	statusCode := 200
-	errorMessage := returnMessage
-
-	return statusCode, returnMessage, errorMessage
+	returnMessage := fmt.Sprintf("Message delivered to topic: %s", topic)
+	return 200, returnMessage, returnMessage
 }
 
 func makeErrorResponse(err error) (int, string, string) {
-	returnMessage := fmt.Sprintf("%s error occured.", err.Error())
-	statusCode := getStatusCodeFromError(err) //400
-	errorMessage := err.Error()
+	returnMessage := fmt.Sprintf("%s error occurred.", err)
 	pkgLogger.Error(returnMessage)
-	return statusCode, returnMessage, errorMessage
+	return getStatusCodeFromError(err), returnMessage, err.Error()
 }
 
 // getStatusCodeFromError parses the error and returns the status so that event gets retried or failed.
 func getStatusCodeFromError(err error) int {
```

Comment:
Did some quick analysis and I can share the following:

These errors were previously considered to be non-retryable but now will be retried:

1. InvalidMessage
2. UnknownTopicOrPartition
3. NotLeaderForPartition
4. RequestTimedOut
5. NotEnoughReplicas


These errors were previously considered to be retryable but now will not be retried:

1. Unknown
2. OffsetOutOfRange
3. ReplicaNotAvailable
4. StaleControllerEpoch
5. RecordListTooLarge
6. InvalidRequiredAcks
7. IllegalGeneration
8. InconsistentGroupProtocol
9. RebalanceInProgress
10. InvalidCommitOffsetSize
11. InvalidTimestamp
12. UnsupportedSASLMechanism
13. IllegalSASLState
14. UnsupportedVersion
15. TopicAlreadyExists
16. InvalidPartitionNumber
17. InvalidReplicationFactor
18. InvalidReplicaAssignment
19. InvalidConfiguration
20. UnsupportedForMessageFormat
21. PolicyViolation
22. OutOfOrderSequenceNumber
23. DuplicateSequenceNumber
24. InvalidProducerEpoch
25. InvalidTransactionState
26. InvalidProducerIDMapping
27. InvalidTransactionTimeout
28. ConcurrentTransactions
29. TransactionCoordinatorFenced
30. TransactionalIDAuthorizationFailed
31. SecurityDisabled
32. BrokerAuthorizationFailed
33. LogDirNotFound
34. SASLAuthenticationFailed
35. UnknownProducerId
36. ReassignmentInProgress
37. DelegationTokenAuthDisabled
38. DelegationTokenNotFound
39. DelegationTokenOwnerMismatch
40. DelegationTokenRequestNotAllowed
41. DelegationTokenAuthorizationFailed
42. DelegationTokenExpired
43. InvalidPrincipalType
44. NonEmptyGroup
45. GroupIdNotFound
46. TopicDeletionDisabled
47. UnsupportedCompressionType
48. StaleBrokerEpoch
49. MemberIDRequired
50. GroupMaxSizeReached
51. FencedInstanceID
52. InvalidRecord 

I cannot find anything really concerning with such a change business-wise, nevertheless it is possible that such a change may disturb some customers, even temporarily.